src/debug_point.c

```diff
@@ -0,0 +1,241 @@
+
+	PG_RETURN_UINT64(debug_point_name_to_id(text_to_cstring(name)));
+}
+
+/*
+ * Wait for the debug point to be released.
+ *
+ * This is handled by first trying to get a shared lock, which will not block
+ * other sessions that try to grab the same lock but will block if an
+ * exclusive lock is already taken, and then release the lock immediately
+ * after.
+ *
+ * This function can decide to block while taking the shared lock or it can
+ * have a retry loop to take the share lock. This retry loop option is useful
+ * in cases where this function gets called from deep down inside a transaction
+ * where interrupts are not being served currently.
+ */
+void
+ts_debug_point_wait(const char *name, bool blocking)
+{
+	DebugPoint point;
+	LockAcquireResult lock_acquire_result pg_attribute_unused();
+	bool lock_release_result pg_attribute_unused();
+
+	debug_point_init(&point, name);
+
+	ereport(DEBUG3, (errmsg("waiting on debug point '%s'", point.name)));
+
+	if (blocking)
+		lock_acquire_result = LockAcquire(&point.tag, ShareLock, true, false);
+	else
+	{
+		/*
+		 * Trying to wait indefinitely here could lead to hangs. The current
+		 * behavior is to retry for retry_count and return with a warning
+		 * if that's crossed.
+		 *
+		 * If required, in future, we could take an additional option to decide
+		 * if the caller wants to retry indefinitely or return with a warning.
+		 * But the current behavior based on the "blocking" argument is ok for
+		 * now.
+		 */
+		unsigned int retry_count = 1000;
+
+		/* try to acquire the lock without waiting. */
+		do
+		{
+			/* try to acquire the lock without waiting. */
+			lock_acquire_result = LockAcquire(&point.tag, ShareLock, true, true);
+
+			if (lock_acquire_result == LOCKACQUIRE_OK)
+				break;
+
+			/* don't dare to take a lock when the proc is exiting! */
+			if (proc_exit_inprogress || ProcDiePending)
+				return;
+
+			if (retry_count == 0)
+			{
+				elog(WARNING, "timeout while acquiring debug point lock");
+				return;
+			}
+			retry_count--;
+
+			/* retry after some time */
+			pg_usleep(100L);
+
+		} while (lock_acquire_result == LOCKACQUIRE_NOT_AVAIL);
+	}
+	Assert(lock_acquire_result == LOCKACQUIRE_OK);
+
+	lock_release_result = LockRelease(&point.tag, ShareLock, true);
+	Assert(lock_release_result);
+
+	ereport(DEBUG3, (errmsg("proceeding after debug point '%s'", point.name)));
+}
+
+/*
+ * Produce an error in case if the debug point is enabled.
+ */
+void
+ts_debug_point_error_injection(const char *name)
+{
+	DebugPoint point;
+	LockAcquireResult lock_acquire_result;
+
+	debug_point_init(&point, name);
+
+	lock_acquire_result = LockAcquire(&point.tag, ExclusiveLock, true, true);
+	if (lock_acquire_result != LOCKACQUIRE_NOT_AVAIL)
+	{
+		/* release/decrement lock count for OK, ALREADY_HELD cases */
+		LockRelease(&point.tag, ExclusiveLock, true);
+	}
+
+	if (lock_acquire_result == LOCKACQUIRE_OK)
```

Comment:
is this check and return based on it correct?

Reply:
There are couple of cases when this can be used. In case if the lock is created inside the same session we will not block and need to release it right away, since it will decrement ref counter.